### DISCUSSION TITLE
Session-level scores

### DISCUSSION THREAD

**Discussion by felixsg** [langfuse/langfuse#2728] [Ideas]
### Describe the feature or potential improvement

At the moment, scores are related to traces and/or observations within a trace but cannot yet be associated with a session. However, there are cases where you might want to provide scores for entire sessions. For instance, in a mental health chatbot, you would want to evaluate whole sessions using different metrics, since individual traces may not provide sufficient context.

**Comment by marcklingen** [comment]
Thanks for sharing, super helpful and something I chatted about with a couple of teams running on langfuse already.

**Comment by gns-fynd** [comment]
Hey @marcklingen 
Is there any updates on langfuse on evaluating the session level? For our use case also we need to evaluate the agents on a session level. So does the langfuse have this support or not.

**Reply by marcklingen** [reply]
Langfuse currently does not have support for adding evaluation scores to a session. Thanks for sharing that this is important to you

**Comment by malikbrh** [comment]
Hi there @marcklingen , any news on supporting scores for sessions? Would be very useful on a project I have for my bachelor degree, I would need to compute mean metrics of multiple traces.

By the way, thanks for all the hard work on the project, you guys rock !

**Reply by marcklingen** [reply]
No update on this yet, thanks for your +1 that this would be useful

**Reply by malikbrh** [reply]
What do you think would be the best solution to achieve mean results of different scores? I thought about building a custom webapp to aggregate scores from Langfuse's API, but before putting multiple hours in this I wanted to know if any other already shipped Langfuse feature would be appropriate for this.

**Reply by marcklingen** [reply]
You can also add this to the run metadata at the end of the run while you might have all evals still available locally

**Comment by marceloribeiro2** [comment]
Hi @marcklingen,

For our project, it is also important to apply scores to all traces based on a sessionId, so that all agents involved receive a common orchestration evaluation.

Currently, I am manually defining the traceIds and storing them in my database because, even when using only langfuse with fetchTraces and iterating to assign scores, the fetch returns too much information, whereas I only need the traceId's to score.

Thank you for the great work and effort put into improving this platform!

**Comment by wbreilly** [comment]
This feature is on Roadmap as "planned". This is mission critical for us. Can you provide a rough estimate for release? 
Thank you

**Reply by marcklingen** [reply]
It is within the eval workstream run by @marliessophie, overall we do not communicate timelines externally as priorities can always change, it is rather a matter of weeks than months

**Reply by fede-bello** [reply]
Are there any news regarding this? Would be really useful

**Reply by marcklingen** [reply]
No, it’s on the list of planned changes for the next weeks though. Will post an update here once this is available

**Comment by tbaker21** [comment]
+1 - I would love to see this feature included in the near future. We have a list of questions our LLM needs to get through in a certain conversation. I'd like metrics on how out of all the conversations the AI has had today, how many of them did the AI get through all of the questions we needed them to ask the user. Thanks.

**Comment by marliessophie** [comment]
Hey there! We have added support for session-level scores in UI and API now: https://langfuse.com/changelog/2025-04-28-session-level-scores. Please let us know in case of any feedback!

**Reply by emranbm** [reply]
That's nice, thanks!  
It'll be much greater if the `LLM-as-a-Judge` evaluators are also supported. e.g., support `Session` as the template variables source with array index indicator support for `jsonPath`.  
After all, thanks for the awesome platform!

**Reply by judsoniv** [reply]
I'm also hoping for extending the LLM-as-Judge feature to support the new Session scores. I could see two features here, both with their own benefits:

1. Aggregate scoring: Pick a score that you're running on traces. For every trace in a session that has a score of that type, sum them and divide by the number of traces that had the score. This would give you a "session-average" score. No _actual_ LLM-as-judge eval would be run on the session as a whole, but you would have an aggregate score (or scores). This would even be good to have as a think that doesn't even have to be set up, just a think you get out of the box for having scores on traces. Kind of an automatic roll-up. This might be the easiest to implement.
2. LLM-eval on a session as a whole. This is more difficult to do both well and in a generalized way that's going to meet everyone's diverse use cases. If I were creating it, I think I'd have a configuration for a repeatable template as well as a prompt, where the eval would first iterate over each trace and put data from that into the provided template, then insert the templated data into the prompt to send to the judge.

**Reply by marcklingen** [reply]
Thanks for sharing! Agree that llm-as-a-judge on sessions is relevant but has its own challenges (there is not timestamp when a session is "finished", thus this needs a new attribute or longer delay), can you create a new thread for this?

**Reply by PE51K** [reply]
Hello! Is there any option to use session-level scores with _langfuse_context_ or _get_client_ API's?

**Reply by aaronleesmith** [reply]
Thanks for this update. Is there any way currently to upload scores for sessions via the API or in any programmatic way? We would like to capture user feedback for while conversations, not individual traces, but have no plans to grant access to the Langfuse UI.

**Reply by PE51K** [reply]
> Hello! Is there any option to use session-level scores with _langfuse_context_ or _get_client_ API's?

Here is the hacky way to set session-level scores via API:
```python
async def create_session_score(
    session_id: str,
    score_name: str,
    score_value: int,
    comment: Optional[str] = None
) -> bool:
    """
    Create a session-level score using Langfuse API
    
    Args:
        session_id: The session ID to associate the score with
        score_name: Name of the score (e.g., "user_satisfaction")  
        score_value: Numeric score value (1-5 for rating)
        comment: Optional comment about the score
        
    Returns:
        bool: True if score was created successfully, False otherwise
    """
    try:
        # Prepare authentication headers
        credentials = f"{settings.langfuse.public_key}:{settings.langfuse.secret_key}"
        encoded_credentials = base64.b64encode(credentials.encode()).decode()
        
        headers = {
            "Content-Type": "application/json",
            "Authorization": f"Basic {encoded_credentials}"
        }
        
        # Prepare score data
        score_data = {
            "sessionId": session_id,
            "name": score_name,
            "value": score_value,
            "dataType": "NUMERIC",
            "comment": comment
        }
        
        # Make HTTP request to Langfuse API
        async with httpx.AsyncClient() as client:
            response = await client.post(
                f"{settings.langfuse.host}/api/public/scores",
                json=score_data,
                headers=headers,
                timeout=10.0
            )
            response.raise_for_status()
            print(f"Successfully created session score: {score_name}={score_value} for session {session_id}")
            return True
            
    except httpx.HTTPStatusError as e:
        print(f"HTTP error creating session score: {e.response.status_code} - {e.response.text}")
        return False

    except Exception as e:
        print(f"Error creating session score: {e}")
        return False
```

`settings.langfuse.host` is your langfuse-web host

**Comment by edebrouwer** [comment]
Thank you for this feature. 
 However, it seems you can only set the score manually for the session. Ideally I would like to compute an average score for the session automatically (or any simple aggregation function of the score). Is that currently possible ?
Thank you

**Reply by edebrouwer** [reply]
like each trace has a score and I want to compute an average session score from the UI

**Reply by mathpinh** [reply]
@marcklingen do you any aupdate on this?

**Comment by judsoniv** [comment]
Exactly On Jun 27, 2025, at 5:08 PM, Edward De Brouwer ***@***.***> wrote:﻿
like each trace has a score and I want to compute an average session score from the UI
